runtime-sdk/modules/evm/src/precompile/erc20.rs

```diff
@@ -0,0 +1,549 @@
+    symbol: u64,
+    decimals: u64,
+    total_supply: u64,
+    balace_of: u64,
+    transfer: u64,
+    transfer_from: u64,
+    approve: u64,
+    allowance: u64,
+    mint: u64,
+    burn: u64,
+}
+
+impl TokenOperationCosts {
+    pub const fn default() -> Self {
+        Self {
+            name: 100,
+            symbol: 100,
+            decimals: 100,
+            total_supply: 500,
+            balace_of: 500,
+            transfer: 1_000,
+            transfer_from: 1_000,
+            approve: 1_500,
+            allowance: 500,
+            mint: 1_000,
+            burn: 1_000,
+        }
+    }
+}
+
+/// Static ERC-20 token implementations should implement this trait to provide
+/// token-specific configuration to the static contract framework.
+pub trait Erc20Token {
+    const GAS_COSTS: TokenOperationCosts;
+
+    const NAME: &str;
+    const SYMBOL: &str;
+    const DECIMALS: u8;
+
+    fn address() -> H160;
+
+    fn total_supply() -> Result<u128, Error>;
+    fn balance_of(account: &H160) -> Result<u128, Error>;
+    fn transfer(sender: &H160, recipient: &H160, amount: u128) -> Result<bool, Error>;
+    fn transfer_from(
+        owner: &H160,
+        spender: &H160,
+        recipient: &H160,
+        amount: u128,
+    ) -> Result<bool, Error>;
+    fn approve(owner: &H160, spender: &H160, amount: u128) -> Result<bool, Error>;
+    fn allowance(owner: &H160, spender: &H160) -> Result<u128, Error>;
+    fn mint(caller: &H160, to: &H160, amount: u128) -> Result<(), Error>;
+    fn burn(caller: &H160, from: &H160, amount: u128) -> Result<(), Error>;
+}
+
+/// Helper trait for simplifying implementations for tokens backed by the SDK's
+/// Accounts module.
+///
+/// The trait presents a much narrower interface than the full [`Erc20Token`]
+/// trait and there is a blanket [`Erc20Token`] impl for any struct that
+/// implements `AccountToken`.
+pub trait AccountToken {
+    type Accounts: AccountsAPI;
+
+    const ADDRESS: H160;
+
+    const GAS_COSTS: TokenOperationCosts;
+
+    const NAME: &str;
+    const SYMBOL: &str;
+    const DECIMALS: u8;
+
+    fn denomination() -> token::Denomination;
+
+    fn is_minting_allowed(caller: &H160, address: &H160) -> Result<bool, Error>;
+    fn is_burning_allowed(caller: &H160, address: &H160) -> Result<bool, Error>;
+}
+
+fn with_allowance<T, F>(owner: &H160, spender: &H160, f: F) -> Result<u128, Error>
+where
+    T: AccountToken,
+    F: FnOnce(Option<u128>) -> Result<u128, Error>,
+{
+    let owner = owner.as_bytes();
+    let spender = spender.as_bytes();
+    CurrentState::with_store(|store| {
+        let store = storage::PrefixStore::new(store, &MODULE_NAME);
+        let allowances = storage::PrefixStore::new(store, &STATE_ALLOWANCES);
+        let denomination = T::denomination().to_string();
+        let allowances = storage::PrefixStore::new(allowances, &denomination);
+        let mut owner = storage::TypedStore::new(storage::PrefixStore::new(allowances, &owner));
```

Comment:
So this is per-denomination and multiple ERC-20 contracts share the same allowances? I guess this makes sense, but also makes it seems like it would be even more suited for the accounts module, even if it is initially only available through the internal API (e.g. no transactions/calls for modifying/using allowances, just API methods that currently only this ERC20 implementation would use).